### PR TITLE
Support regex for application names

### PR DESCRIPTION
**Pull request by Frederick888**
I'd like to map left Ctrl as Esc when pressed alone unless I'm in
Minecraft. Since Minecraft adds its version number to the window class
name, e.g. `Minecraft* 1.19.2` (and I think I have an `*` cos I use
mods), it's nice to be able to use regular expressions.

Thanks for this great application btw!

### CODE REVIEW

File: src/config/application.rs [hunk 1]
```diff
@@ -1,13 +1,63 @@
...
+
+impl FromStr for ApplicationMatcher {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let first_char = s.chars().next();
+        match first_char {
+            None => Err(anyhow!("Empty application name")),
+            Some('~') if s.len() == 1 => Err(anyhow!("Empty regular expression for application name")),
+            Some('~') => Ok(ApplicationMatcher::Regex(Regex::new(&s[1..])?)),
```
Comment (k0kubun) [inline src/config/application.rs:40]: I'm not used to the syntax of declaring a regular expression by starting it with `~`. How about wrapping a regex with `/`s [like CircleCI](https://circleci.com/docs/workflows#using-regular-expressions-to-filter-tags-and-branches), e.g. `/regex/`?

Also, it'd be nice to test parsing it at `src/config/tests.rs`.

Comment (Frederick888) [reply src/config/application.rs:40]: Thanks for the review! Just two quick questions before I move forward:

1. Is there a straightforward way to unescape PCRE regular expressions and create `regex::Regex`? I'm happy to write a simple unescape function but just want to confirm I'm not missing anything
2. Do you also want it to support modifiers, e.g. `/^minecraft/i` for case-insensitive matching?

Comment (k0kubun) [reply src/config/application.rs:40]: > Is there a straightforward way to unescape PCRE regular expressions and create regex::Regex? I'm happy to write a simple unescape function but just want to confirm I'm not missing anything

I don't know of one. I was just thinking about checking if it starts and ends with a `/` and removing them.

> Do you also want it to support modifiers, e.g. `/^minecraft/i` for case-insensitive matching?

No, I don't. For simplicity, I prefer starting from supporting no flags. I don't think the application name would fluctuate like minecraft, MineCraft, or Minecraft either.

Thanks for asking :slightly_smiling_face:

Comment (Frederick888) [reply src/config/application.rs]: Pushed. Could you have a look again?

**Review by k0kubun — APPROVED** [review]
Thank you!

**Pull request merged**